Explicit `CachedFuzzySearchable<T>` wrapper

LuizZak opened the pull request:
As discussed in #2, this is an explicit structure that wraps over an existing, non-cacheable `FuzzySearchable`, providing a thin layer that allows for caching of the tokenized searchable string.

This version has a few advantages that I'm planning on checking and proving related to resilience against sharing cache across multiple value-type `FuzzySearchable` implementers, since the implementer is copied into a private, constant field within the `CachedFuzzySearchable<T>` wrapper.

viktorasl commented on Source/FuzzySearch.swift:
```diff
@@ -46,13 +62,53 @@ public struct FuzzySearchResult {
 public protocol FuzzySearchable {
     var fuzzyStringToMatch: String { get }
+    
+    func fuzzyTokenized() -> FuzzyTokens
```
I think `fuzzyTokenized` method could be private/internal unless we would like it to be overridable.

viktorasl commented on Tests/FuzzySearchTests.swift:
```diff
@@ -88,4 +124,15 @@ class FuzzySearchTests: XCTestCase {
+        let spanishWordsCached = spanishWords.map(CachedFuzzySearchable.init)
+        
+        measure {
+            _ = spanishWordsCached.fuzzyMatch("la sart")
```
Fuzzy searching whole "la sart" won't show better performance right away, because it will tokenize each word only once, what you should do is:
```swift
measure {
            _ = spanishWordsCached.fuzzyMatch("l")
            _ = spanishWordsCached.fuzzyMatch("la")
            _ = spanishWordsCached.fuzzyMatch("la ")
            _ = spanishWordsCached.fuzzyMatch("la s")
            _ = spanishWordsCached.fuzzyMatch("la sa")
            _ = spanishWordsCached.fuzzyMatch("la sar")
            _ = spanishWordsCached.fuzzyMatch("la sart")
 }
```
something like that. Change the implementation in `testSpeedOfFuzzySearchFor1000SpanishWords` as well.

viktorasl commented:
Everything looks great! Let me know which PR you prefer.

LuizZak commented:
@viktorasl Cool! I think this seems like the preferred cleaner way of dealing with caching, it's decoupled from the base protocol and allows for discarding of caching, in case it's heavyweight to keep around.

You can merge this version in!

viktorasl commented:
Could you take a look to those 2 comments?

LuizZak commented:
Sorry, can't seem to find them, where?

viktorasl commented:
Sorry, didn't see that I didn't submit review and it was pending

LuizZak commented:
Addressed review, I've had to be slightly creative with the public/private protocol implementations to hide away `fuzzyTokenized` and still make it transparently available to `CachedFuzzySearchable.fuzzyMatch`. More info is available in a comment down at that method's implementation.
